0005-ordered-quantification.md

```diff
@@ -0,0 +1,626 @@
+  | Ident [":" Type] "<-" Expression
+  | CasePatternLocal "<-" Expression
+
+QuantifierDomain ::=
+  QuantifierVarDecl 
+  { "," QuantifierVarDecl }
+  [ "|" Expression ]
+```
+
+Note that this concept and the new `<-` syntax applies to any use of quantifier domains, even though ordering and multiplicity
+is irrelevant for all current uses: the semantics of `[i]set` and `map` comprehensions, `forall` and `exists` expressions, 
+and `forall` statements all do not depend on the order of quantification and semantically ignore duplicates.
+Importantly, these extensions to the syntax and semantics of quantifier domains are all fully backwards compatible.
+They do offer a more efficient expression of common uses of these existing features, however: `set Some(x) <- s` is equivalent to
+`set x | x in s && x.Some? :: x.value`, for example.
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x <- [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x <- myDoubleDeckerSeq, y <- x | y != 0 {
+  Process(y);
+}
+```
+
+The high-level syntax for a `foreach` loop reuses several existing concepts:
+
+```
+ForeachLoop ::=
+  "foreach"
+  QuantifierDomain
+  LoopSpec
+  Body
+```
+
+More specifically, this expands to:
+
+```
+ForeachLoop ::=
+  "foreach" 
+  QuantifierVarDecl { "," QuantifierVarDecl } 
+  [ "|" Expression ]
+  { InvariantDecl | ModifiesDecl | DecreasesDecl }
+  [ BlockStmt ]
+```
+
+A `foreach` loop closely resembles a `forall` statement, the key difference being that a `forall`
+statement executes its body once for every binding of quantified values simultaneously, whereas a `foreach` loop
+executes its body once for each binding of quantified values in sequence, one at a time.
+
+Similarly to set comprehensions or assign-such-that statements, the domain of a `foreach` loop
+must be enumerable. The following loop would produce the error `"the domain of a foreach loop must be enumerable,
+but Dafny's heuristics can't figure out how to produce an enumeration for 'x'"`.
+
+```dafny
+foreach x: real | 0.0 <= x < 1.0 {  // Error: ...
+  ...
+}
+```
+
+The quantifier domain is allowed to be potentially infinite if the loop is used in a compiled context and an explicit `decreases` clause is provided.
+`decreases *` is permitted, in which the `foreach` loop may never terminate. Any other `decreases` clause can be provided
+to ensure the loop terminates even if the domain is potentially infinite. For example, the following (very slow) example collects
+five arbitrary primes (assuming that Dafny can figure out how to enumerate the `allPrimes` infinite set):
```

Comment:
It would also need to know that it's not empty, right?

Reply:
Worse, it would have to know there are at least five elements. Fortunately the code is sound, but my statement about how it behaves is wrong - it should say that it collects AT MOST five arbitrary primes. I'll fix that.